### DISCUSSION TITLE
Compute shader branching depending on buffer value

### DISCUSSION THREAD

**Discussion by h-a-n-n-e-s** [gpuweb/gpuweb#4374] [Q&A] [answered]
I would like to reuse a pipeline, bind group, etc. for two computation stages using the same shader code. What I do might be weird, but I would still like to understand what's going on.
There is a shader
```wgsl
@group(0) @binding(0) var<uniform> stage: u32;

fn main(...) {
  if stage == 0 {
    doOneThing();
  }
  if stage == 1 {
    doAnotherThing();
  }
}
```

Then I use

```js
device.queue.writeBuffer(stageBuffer, 0, new Uint32Array([0])); // set stage = 0
{
  const commandEncoder = device.createCommandEncoder();
  const passEncoder = commandEncoder.beginComputePass();
  ...
  passEncoder.dispatchWorkgroups(...)
  device.queue.submit([commandEncoder.finish()]);
}

device.queue.writeBuffer(stageBuffer, 0, new Uint32Array([1])); // set stage = 1
{
  // same as above
  const commandEncoder = device.createCommandEncoder();
  const passEncoder = commandEncoder.beginComputePass();
  ...
  passEncoder.dispatchWorkgroups(...)
  device.queue.submit([commandEncoder.finish()]);
}
```

The problem is that for the second submission in the shader the `stage` is still `0`, even though I set it to `1` before. Why?

If I first set `stage = 1` then stage 1 is executed. It is always the second submission where the new stageBuffer value is ignored in the shader, but the debugger is showing the new value. What am I missing here?

**Comment by Kangz** [comment]
At first glance the code looks like it might work (assuming all the redacted out code is correct). So this could be an implementation bug. A bug filed on the implementation as well as a minimal repro would help.

**Reply by h-a-n-n-e-s** [reply, marked as answer]
Turns out I relied on [webgpu-devtools](https://github.com/takahirox/webgpu-devtools) which does not seem to properly update the buffer info in this case. : /

Everything works as expected!
Sorry for the disturbance!
